### DISCUSSION TITLE
Tricks to get faster function calls?

### DISCUSSION THREAD

**Discussion by fasaxc** [micropython/micropython#15483] [Using MicroPython]
I'm writing an ISR using viper and I need to poke at some registers from the ISR. The nice way to do that is to call C-implemented API functions (for example `PWM.duty_u16()`) but it seems the function call overhead is pretty big: 8-10us. If I inline the function (just directly set registers in the ISR then it's well under 1us).

I was wondering if there are any tricks to get faster function calls so I can avoid reinventing the wheel?

**Comment by bixb922** [comment]
Maybe you already tried this, but on some microcontrollers you can change hardware device registers of the PWM directly. With Viper, ```ptr32(memory_address)[0]``` allows to read /write those registers directly. That can be much faster than a function call. The microcontroller's technical reference manual normally shows register addresses and bit assignments.

**Reply by fasaxc** [reply]
Yes, that's what I've been doing but it means that I need to figure out the registers, which is awkward on the rp2040 because resources are dynamically allocated. For example, if I ask for a PIO state machine, I might get SM0 or I might get SM1 depending on whether one has already been allocated. I'm writing library code so that makes it awkward.
